deps/spidershim/src/v8functiontemplate.cc

```diff
@@ -0,0 +1,315 @@
+  } else {
+    js::SetReservedSlot(templateObj, CallbackSlot, JS::UndefinedValue());
+  }
+
+  if (*data) {
+    js::SetReservedSlot(templateObj, DataSlot, *GetValue(data));
+  } else {
+    js::SetReservedSlot(templateObj, DataSlot, JS::UndefinedValue());
+  }
+}
+
+bool
+TemplateIsInstantiated(JSObject* templateObj)
+{
+  assert(js::GetReservedSlot(templateObj, InstanceSlot).isObject() ||
+         js::GetReservedSlot(templateObj, InstanceSlot).isUndefined());
+  return js::GetReservedSlot(templateObj, InstanceSlot).isObject();
+}
+} // anonymous namespace
+
+namespace v8 {
+
+// static
+Local<FunctionTemplate> FunctionTemplate::New(Isolate* isolate,
+                                              FunctionCallback callback,
+                                              Handle<Value> data,
+                                              Handle<Signature> signature,
+                                              int length)
+{
+  Local<Template> templ = Template::New(isolate, &functionTemplateClass);
+  if (*templ) {
+    return Local<FunctionTemplate>();
+  }
+
+  JSContext* cx = JSContextFromIsolate(isolate);
+  JS::RootedObject obj(cx, GetObject(*templ));
+  assert(obj);
+
+  SetCallbackAndData(obj, callback, data);
+
+  // TODO: figure out what should happen with signature.  What is it supposed to
+  // do?
+  js::SetReservedSlot(obj, LengthSlot, JS::Int32Value(length));
+
+  return Local<FunctionTemplate>(static_cast<FunctionTemplate*>(*templ));
+}
+
+Local<Function> FunctionTemplate::GetFunction()
+{
+  MaybeLocal<Function> maybeFunc =
+    GetFunction(Isolate::GetCurrent()->GetCurrentContext());
+  return maybeFunc.FromMaybe(Local<Function>());
+}
+
+MaybeLocal<Function> FunctionTemplate::GetFunction(Local<Context> context)
+{
+  JSContext* cx = JSContextFromContext(*context);
+  JS::RootedObject obj(cx, GetObject(this));
+  assert(obj);
+
+  // XXXbz V8 uses a temporary HandleScope here, but I _think_ that's some sort
+  // of performance optimization...
+
+  if (TemplateIsInstantiated(obj)) {
+    JS::RootedValue existingInstance(cx,
+                                     js::GetReservedSlot(obj, InstanceSlot));
+    assert(existingInstance.isObject());
+    return internal::Local<Function>::New(context->GetIsolate(),
+                                          existingInstance);
+  }
+
+  JS::RootedValue callbackVal(cx, js::GetReservedSlot(obj, CallbackSlot));
+  // XXXbz the .get() here is working around
+  // https://bugzilla.mozilla.org/show_bug.cgi?id=1275082
+  auto callback = reinterpret_cast<FunctionCallback>(callbackVal.get().toPrivate());
+
+  JS::RootedValue dataVal(cx, js::GetReservedSlot(obj, DataSlot));
+  // XXXbz is this a sane thing to do?  Or do I need to register this thing with
+  // some HandleScope?  Seems like I need to do the latter if the callee has
+  // this thing outlive the Rooted, right?
+  Local<Value> dataV8Val = GetV8Value(dataVal.address());
```

Comment:
The `conversions.h` routines can only be used for stuff currently registered with a `HandleScope`.  What you want to do here instead is something like:

```
Local<Value> dataV8Val = internal::Local<Value>::New(isolate, js::getReservedSlot(...));
```

This way you won't need the `dataVal` root either.